File: taskmanager.py
import sqlite3
import os
from PyQt6.QtCore import QObject, pyqtSignal, QDateTime

class TaskManager(QObject):
    tasksChanged = pyqtSignal()

    def __init__(self, db_path="~/.tooodo/db/tooodo.db", parent=None):
        super().__init__(parent)
        self.db_path = os.path.expanduser(db_path)
        self.ensure_directory_exists()
        self.create_tasks_table()

    def ensure_directory_exists(self):
        directory = os.path.dirname(self.db_path)
        if not os.path.exists(directory):
            os.makedirs(directory, exist_ok=True)

    def create_tasks_table(self):
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS tooodo (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                title TEXT NOT NULL,
                description TEXT,
                due_date DATETIME,
                priority INT,
                completed BOOL
            )
        ''')
        conn.commit()
        conn.close()

    def add_task(self, title, description, due_date):
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()

        cursor.execute('''
            INSERT INTO tooodo (title, description, due_date) VALUES (?, ?, ?)
        ''', (title, description, due_date.toString("yyyy-MM-dd HH:mm:ss")))
        conn.commit()
        conn.close()
        self.tasksChanged.emit()

    def get_tasks(self):
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        cursor.execute('SELECT * FROM tooodo')
        tasks = cursor.fetchall()
        conn.close()
        return tasks

    def complete_task(self, task_id):
        self.update_task(task_id, {"completed": 1})

    def update_task(self, task_id, data):
        update_str = ", ".join(["{0} = ?".format(k) for k in data])

        values = list(data.values())
        values.append(task_id)

        if not update_str:
            return

        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        cursor.execute(f'UPDATE tooodo SET {update_str} WHERE id = ?', values)
        conn.commit()
        conn.close()
        self.tasksChanged.emit()

File: src/taskmanager/taskmanager.cpp
#include "taskmanager.h"
#include "taskitem/taskitem.h"
#include <QDir>
#include <QxOrm.h>
#include <QSqlDatabase>
#include <QSqlQuery>

TaskManager::TaskManager(QObject *parent)
    : QObject(parent)
{
    db_path = QDir::homePath() + "/.tooodo/db";
    ensureDirectoryExists();
    createTasksTable();
    // printTableSchema();
}

void TaskManager::ensureDirectoryExists()
{
    QDir dir(db_path);

    if (!dir.exists()) {
        dir.mkpath(".");
    }
}

void TaskManager::createTasksTable()
{
    QString db_file = db_path + "/tooodo.db";

    qx::QxSqlDatabase::getSingleton()->setDriverName("QSQLITE");
    qx::QxSqlDatabase::getSingleton()->setDatabaseName(db_file);
    qx::QxSqlDatabase::getSingleton()->getDatabase();

    qx::dao::create_table<TaskItem>();
}

void TaskManager::addTask(TaskItem *task)
{
    qx::dao::insert(task);
    emit tasksChanged();
}

void TaskManager::updateTask(TaskItem *task)
{
    qx::dao::save(task);
    emit tasksChanged();
}

QList<TaskItem*> TaskManager::getTasks()
{
    QList<TaskItem*> taskList;
    qx::dao::fetch_all(taskList);
    return taskList;
}

void TaskManager::printTableSchema()
{
    QString db_file = db_path + "/tooodo.db";

    QSqlDatabase db = QSqlDatabase::addDatabase("QSQLITE", "schema_connection");
    db.setDatabaseName(db_file);

    if (!db.open())
    {
        qDebug() << "Failed to open database:" << db.lastError().text();
        return;
    }

    QString tableName = "TaskItem";
    QSqlQuery query(db);
    if (query.exec("SELECT * from " + tableName + ";"))
    {
        while (query.next())
       {
           int id = query.value("id").toInt();
           QString title = query.value("title").toString();
           QString description = query.value("description").toString();
           QDateTime dueDate = query.value("due_date").toDateTime();
           int priority = query.value("priority").toInt();
           bool completed = query.value("completed").toBool();

           qDebug() << "ID:" << id
                    << "Title:" << title;
       }
    }
    else
    {
        qDebug() << "Failed to retrieve table schema:" << query.lastError().text();
    }

    db.close();
}
